File: packages/core/src/crypto/encrypt.ts
/**
 * 文件名加密
 * TS 重写 encrypt_epub.py
 */
import { createHash } from 'node:crypto';
import path from 'node:path';
import { EpubParser, type ManifestItem } from '../epub/parser.js';
import { EpubWriter } from '../epub/writer.js';
import { logger } from '../utils/logger.js';
import type JSZip from 'jszip';

export interface EncryptOptions {
  /** 输出路径 */
  outputPath: string;
}

export interface EncryptResult {
  success: boolean;
  filesEncrypted: number;
  mapping: Map<string, string>;
}

/** Media type categories and their OEBPS sub-directories */
type FileCategory = 'text' | 'css' | 'image' | 'font' | 'audio' | 'video' | 'other';

const CATEGORY_DIR: Record<FileCategory, string> = {
  text: 'Text',
  css: 'Styles',
  image: 'Images',
  font: 'Fonts',
  audio: 'Audio',
  video: 'Video',
  other: 'Misc',
};

/**
 * MD5(id) → 二进制 → 1→'*', 0→':' → 作为新文件名
 */
export function generateEncryptedName(id: string): string {
  const md5 = createHash('md5').update(id).digest('hex');
  const binary = BigInt(`0x${md5}`).toString(2).padStart(128, '0');
  return binary.replace(/1/g, '*').replace(/0/g, ':');
}

/** Classify a manifest item into a file category */
function classifyItem(item: ManifestItem): FileCategory {
  const mt = item.mediaType;
  const href = item.href.toLowerCase();
  if (mt === 'application/xhtml+xml') return 'text';
  if (mt === 'text/css') return 'css';
  if (mt.startsWith('image/')) return 'image';
  if (mt.startsWith('font/') || href.endsWith('.ttf') || href.endsWith('.otf') || href.endsWith('.woff')) return 'font';
  if (mt.startsWith('audio/')) return 'audio';
  if (mt.startsWith('video/')) return 'video';
  return 'other';
}

/** Build the encrypted filename for a manifest item.
 *  ID format: typically "filename.ext" — we split on first '.' to get the name part for hashing. */
function buildEncryptedFilename(id: string, href: string): string {
  const idName = id.split('.')[0];
  const lastDot = href.lastIndexOf('.');
  const ext = lastDot >= 0 ? href.slice(lastDot + 1).toLowerCase() : '';
  const filenameWithoutExt = path.basename(href, path.extname(href));

  let slimSuffix = '';
  let hashInput = idName;
  if (filenameWithoutExt.endsWith('slim') || idName.endsWith('slim')) {
    slimSuffix = '~slim';
    hashInput = idName
      .replace(/~slim$/, '')
      .replace(/-slim$/, '')
      .replace(/_slim$/, '')
      .replace(/slim$/, '');
  }

  const encrypted = generateEncryptedName(hashInput);
  return `_${encrypted}${slimSuffix}.${ext}`;
}

/**
 * Resolve a relative href against a base file path to get the book-path (zip path).
 * e.g. resolveBookPath('../Images/cover.jpg', 'OEBPS/Text/chapter1.xhtml')
 *      → 'OEBPS/Images/cover.jpg'
 */
function resolveBookPath(href: string, basePath: string): string {
  const parts = href.split(/[\\/]/);
  const baseParts = basePath.split(/[\\/]/);
  baseParts.pop(); // remove filename

  let backSteps = 0;
  while (parts.length > 0 && parts[0] === '..') {
    backSteps++;
    parts.shift();
  }
  // Remove '.' parts
  while (parts.length > 0 && parts[0] === '.') {
    parts.shift();
  }

  if (backSteps >= baseParts.length) {
    return parts.join('/');
  }
  const base = baseParts.slice(0, baseParts.length - backSteps);
  return [...base, ...parts].join('/');
}

/**
 * 对 EPUB 文件进行文件名加密
 */
export async function encryptEpub(
  epubPath: string,
  options: EncryptOptions,
): Promise<EncryptResult> {
  const parsed = await EpubParser.fromFile(epubPath);
  const { zip, manifest, spine, opfPath, opfDir, rawOpf } = parsed;

  // Detect NCX toc id from spine
  const opfXml = rawOpf;
  const tocIdMatch = opfXml.match(/<spine[^>]*toc="([^"]+)"/);
  const tocId = tocIdMatch ? tocIdMatch[1] : '';

  // Build categorized lists and path mapping
  // pathMap: old bookpath → { newBasename, category }
  const pathMap = new Map<string, { newBasename: string; category: FileCategory }>();
  // hrefToNew: old opf-relative href → new encrypted basename (for toc/reference rewriting)
  const hrefToNew = new Map<string, string>();
  const usedNames = new Map<FileCategory, Set<string>>();
  for (const cat of Object.keys(CATEGORY_DIR) as FileCategory[]) {
    usedNames.set(cat, new Set());
  }

  let tocBookPath = '';

  for (const [id, item] of manifest) {
    // Skip NCX toc file — handle separately
    if (tocId && id === tocId) {
      const tocHref = item.href;
      tocBookPath = opfDir ? `${opfDir}/${tocHref}` : tocHref;
      continue;
    }

    const category = classifyItem(item);
    let newName = buildEncryptedFilename(id, item.href);

    // Ensure uniqueness within category
    const nameSet = usedNames.get(category)!;
    if (nameSet.has(newName)) {
      const extIdx = newName.lastIndexOf('.');
      const base = newName.slice(0, extIdx);
      const ext = newName.slice(extIdx);
      let counter = 1;
      while (nameSet.has(`${base}_${counter}${ext}`)) counter++;
      newName = `${base}_${counter}${ext}`;
    }
    nameSet.add(newName);

    const bookPath = opfDir ? `${opfDir}/${item.href}` : item.href;
    pathMap.set(bookPath, { newBasename: newName, category });
    hrefToNew.set(item.href, newName);

    logger.debug(`encrypt: ${id}:${item.href} → ${newName}`);
  }

  // Build new zip
  const JSZipModule = (await import('jszip')).default;
  const newZip: JSZip = new JSZipModule();

  // mimetype (JSZip typing expects 'DEFLATE'|'STORE' but its type definition is narrow)
  newZip.file('mimetype', 'application/epub+zip');

  // container.xml — point to OEBPS/content.opf
  const containerXml = await zip.file('META-INF/container.xml')?.async('text');
  if (containerXml) {
    const updatedContainer = containerXml.replace(
      /<rootfile[^>]*media-type="application\/oebps-[^>]*\/>/,
      '<rootfile full-path="OEBPS/content.opf" media-type="application/oebps-package+xml"/>',
    );
    newZip.file('META-INF/container.xml', updatedContainer);
  }

  // Copy & rewrite content files
  for (const [bookPath, { newBasename, category }] of pathMap) {
    const file = zip.file(bookPath);
    if (!file) continue;

    const dir = CATEGORY_DIR[category];

    if (category === 'text') {
      let text = await file.async('text');
      text = rewriteXhtml(text, bookPath, pathMap);
      newZip.file(`OEBPS/${dir}/${newBasename}`, text);
    } else if (category === 'css') {
      let css = await file.async('text');
      css = rewriteCss(css, bookPath, pathMap);
      newZip.file(`OEBPS/${dir}/${newBasename}`, css);
    } else {
      const data = await file.async('nodebuffer');
      newZip.file(`OEBPS/${dir}/${newBasename}`, data);
    }
  }

  // NCX toc file
  if (tocBookPath) {
    const tocFile = zip.file(tocBookPath);
    if (tocFile) {
      let toc = await tocFile.async('text');
      toc = rewriteNcx(toc, tocBookPath, hrefToNew);
      newZip.file('OEBPS/toc.ncx', toc);
    }
  }

  // Rebuild OPF
  const newOpf = rebuildOpf(rawOpf, manifest, tocId, pathMap, opfDir);
  newZip.file('OEBPS/content.opf', newOpf);

  // Write output
  await EpubWriter.write(newZip, { outputPath: options.outputPath });

  logger.info(`Encrypted ${pathMap.size} files → ${options.outputPath}`);

  return {
    success: true,
    filesEncrypted: pathMap.size,
    mapping: new Map([...pathMap].map(([k, v]) => [k, `OEBPS/${CATEGORY_DIR[v.category]}/${v.newBasename}`])),
  };
}

// --------------- Rewrite helpers ---------------

/** Rewrite href/src/xlink:href references in XHTML content */
function rewriteXhtml(
  text: string,
  xhtmlBookPath: string,
  pathMap: Map<string, { newBasename: string; category: FileCategory }>,
): string {
  // Rewrite href="..." attributes
  text = text.replace(
    /(<[^>]*href=(["']))(.*?)(\2[^>]*>)/g,
    (match, prefix, _q, href, suffix) => {
      return rewriteHtmlAttr(match, prefix, href, suffix, xhtmlBookPath, pathMap, true);
    },
  );

  // Rewrite src="..." attributes
  text = text.replace(
    /(<[^>]*\ssrc=(["']))(.*?)(\2[^>]*>)/g,
    (match, prefix, _q, href, suffix) => {
      return rewriteHtmlAttr(match, prefix, href, suffix, xhtmlBookPath, pathMap, false);
    },
  );

  // Rewrite xlink:href="..." attributes (SVG images etc.)
  text = text.replace(
    /(<[^>]*xlink:href=(["']))(.*?)(\2[^>]*>)/g,
    (match, prefix, _q, href, suffix) => {
      return rewriteHtmlAttr(match, prefix, href, suffix, xhtmlBookPath, pathMap, false);
    },
  );

  // Rewrite poster="..." attributes
  text = text.replace(
    /(<[^>]*\sposter=(["']))(.*?)(\2[^>]*>)/g,
    (match, prefix, _q, href, suffix) => {
      return rewriteHtmlAttr(match, prefix, href, suffix, xhtmlBookPath, pathMap, false);
    },
  );

  // Rewrite url() in inline styles
  text = text.replace(
    /(url\(["']?)(.*?)(["']?\))/g,
    (match, pre, url, post) => {
      return rewriteUrl(match, pre, url.trim(), post, xhtmlBookPath, pathMap);
    },
  );

  return text;
}

/** Rewrite a single HTML attribute value (href or src) */
function rewriteHtmlAttr(
  original: string,
  prefix: string,
  rawHref: string,
  suffix: string,
  basePath: string,
  pathMap: Map<string, { newBasename: string; category: FileCategory }>,
  isHref: boolean,
): string {
  const href = decodeURIComponent(rawHref).trim();
  if (!href || href.startsWith('http://') || href.startsWith('https://') || href.startsWith('data:') || href.startsWith('mailto:')) {
    return original;
  }

  let hrefBase = href;
  let fragment = '';
  const hashIdx = href.indexOf('#');
  if (hashIdx >= 0) {
    hrefBase = href.slice(0, hashIdx);
    fragment = href.slice(hashIdx);
  }

  if (!hrefBase) return original; // fragment-only link

  const bookPath = resolveBookPath(hrefBase, basePath);
  const entry = pathMap.get(bookPath);
  if (!entry) return original;

  const dir = CATEGORY_DIR[entry.category];
  const relativePath = `../${dir}/${entry.newBasename}`;

  // For CSS link tags, rewrite the full tag
  if (isHref && hrefBase.toLowerCase().endsWith('.css')) {
    return `<link href="../Styles/${entry.newBasename}" type="text/css" rel="stylesheet"/>`;
  }

  return prefix + relativePath + fragment + suffix;
}

/** Rewrite url() references in CSS or inline styles */
function rewriteUrl(
  original: string,
  pre: string,
  url: string,
  post: string,
  basePath: string,
  pathMap: Map<string, { newBasename: string; category: FileCategory }>,
): string {
  const decoded = decodeURIComponent(url).trim();
  if (!decoded || decoded.startsWith('http://') || decoded.startsWith('https://') || decoded.startsWith('data:')) {
    return original;
  }
  const bookPath = resolveBookPath(decoded, basePath);
  const entry = pathMap.get(bookPath);
  if (!entry) return original;

  const dir = CATEGORY_DIR[entry.category];
  return `${pre}../${dir}/${entry.newBasename}${post}`;
}

/** Rewrite CSS file content — @import and url() references */
function rewriteCss(
  css: string,
  cssBookPath: string,
  pathMap: Map<string, { newBasename: string; category: FileCategory }>,
): string {
  // Rewrite @import
  css = css.replace(
    /@import\s+(?:(?:(["'])(.*?)\1)|(?:url\(["']?(.*?)["']?\)))/g,
    (match, _q, hrefQuoted, hrefUrl) => {
      const href = (hrefQuoted || hrefUrl || '').trim();
      if (!href.toLowerCase().endsWith('.css')) return match;
      const decoded = decodeURIComponent(href);
      const bookPath = resolveBookPath(decoded, cssBookPath);
      const entry = pathMap.get(bookPath);
      if (!entry) return match;
      return hrefQuoted !== undefined
        ? `@import "${entry.newBasename}"`
        : `@import url("${entry.newBasename}")`;
    },
  );

  // Rewrite url()
  css = css.replace(
    /(url\(["']?)(.*?)(["']?\))/g,
    (match, pre, url, post) => {
      return rewriteUrl(match, pre, url.trim(), post, cssBookPath, pathMap);
    },
  );

  return css;
}

/** Rewrite NCX toc content references */
function rewriteNcx(
  toc: string,
  tocBookPath: string,
  hrefToNew: Map<string, string>,
): string {
  return toc.replace(
    /src=(["'])(.*?)\1/g,
    (match, quote, rawHref) => {
      const href = decodeURIComponent(rawHref).trim();
      const hashIdx = href.indexOf('#');
      const hrefBase = hashIdx >= 0 ? href.slice(0, hashIdx) : href;
      const fragment = hashIdx >= 0 ? href.slice(hashIdx) : '';

      // NCX src values are relative to the toc file; resolve to bookpath for matching.
      // We build a synthetic OPF reference path based on the toc's directory to resolve
      // opf-relative hrefs from hrefToNew against the same directory tree.
      const tocDir = tocBookPath.slice(0, tocBookPath.lastIndexOf('/') + 1);
      for (const [oldHref, newName] of hrefToNew) {
        const oldBookPath = resolveBookPath(oldHref, tocDir + '_opf_ref');
        const resolvedHref = resolveBookPath(hrefBase, tocBookPath);
        if (oldBookPath === resolvedHref) {
          return `src="Text/${newName}${fragment}"`;
        }
      }
      return match;
    },
  );
}

/** Rebuild OPF with updated manifest hrefs */
function rebuildOpf(
  opf: string,
  manifest: Map<string, ManifestItem>,
  tocId: string,
  pathMap: Map<string, { newBasename: string; category: FileCategory }>,
  opfDir: string,
): string {
  // Build new manifest block
  let manifestText = '<manifest>';

  for (const [id, item] of manifest) {
    const props = item.properties ? ` properties="${item.properties}"` : '';

    if (tocId && id === tocId) {
      manifestText += `\n    <item id="${id}" href="toc.ncx" media-type="application/x-dtbncx+xml"/>`;
      continue;
    }

    const bookPath = opfDir ? `${opfDir}/${item.href}` : item.href;
    const entry = pathMap.get(bookPath);
    if (!entry) {
      // Keep original href for unmapped items
      manifestText += `\n    <item id="${id}" href="${item.href}" media-type="${item.mediaType}"${props}/>`;
      continue;
    }

    const dir = CATEGORY_DIR[entry.category];
    manifestText += `\n    <item id="${id}" href="${dir}/${entry.newBasename}" media-type="${item.mediaType}"${props}/>`;
  }

  manifestText += '\n  </manifest>';

  // Replace manifest in OPF
  let newOpf = opf.replace(/(?:<manifest[\s\S]*?<\/manifest>)/i, manifestText);

  // Update <reference> hrefs in guide section
  newOpf = newOpf.replace(
    /(<reference[^>]*href=(["']))(.*?)(\2[^>]*\/>)/g,
    (match, prefix, _q, href, suffix) => {
      const decoded = decodeURIComponent(href).trim();
      if (decoded.endsWith('.ncx')) return match;

      // Strip leading relative path segments (./, ../, /)
      let cleaned = decoded;
      if (cleaned.startsWith('/')) cleaned = cleaned.slice(1);
      while (cleaned.startsWith('../')) cleaned = cleaned.slice(3);
      if (cleaned.startsWith('./')) cleaned = cleaned.slice(2);

      // Find in pathMap via opf-relative resolution
      const bookPath = opfDir ? `${opfDir}/${cleaned}` : cleaned;
      const entry = pathMap.get(bookPath);
      if (!entry) return match;
      return `${prefix}Text/${entry.newBasename}${suffix}`;
    },
  );

  return newOpf;
}

File: crates/epub-core/src/crypto/encrypt.rs
//! Filename encryption for EPUB files.
//!
//! Port of `packages/core/src/crypto/encrypt.ts`.
//!
//! Algorithm: MD5(id) → 128-bit binary → `1` → `*`, `0` → `:` → encrypted filename.

use std::collections::{HashMap, HashSet};

use md5::{Digest, Md5};

use crate::epub::parser::ManifestItem;

// ─── Types ────────────────────────────────────────────────────────────────────

/// Result of an encryption operation.
#[derive(Debug)]
pub struct EncryptResult {
    pub success: bool,
    pub files_encrypted: usize,
    /// old bookpath → new bookpath
    pub mapping: HashMap<String, String>,
}

// ─── File categories ──────────────────────────────────────────────────────────

#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
pub enum FileCategory {
    Text,
    Css,
    Image,
    Font,
    Audio,
    Video,
    Other,
}

impl FileCategory {
    /// OEBPS sub-directory name for this category.
    pub fn dir_name(self) -> &'static str {
        match self {
            Self::Text => "Text",
            Self::Css => "Styles",
            Self::Image => "Images",
            Self::Font => "Fonts",
            Self::Audio => "Audio",
            Self::Video => "Video",
            Self::Other => "Misc",
        }
    }
}

/// Classify a manifest item into a file category.
pub fn classify_item(item: &ManifestItem) -> FileCategory {
    let mt = &item.media_type;
    let href = item.href.to_lowercase();
    if mt == "application/xhtml+xml" {
        FileCategory::Text
    } else if mt == "text/css" {
        FileCategory::Css
    } else if mt.starts_with("image/") {
        FileCategory::Image
    } else if mt.starts_with("font/")
        || href.ends_with(".ttf")
        || href.ends_with(".otf")
        || href.ends_with(".woff")
    {
        FileCategory::Font
    } else if mt.starts_with("audio/") {
        FileCategory::Audio
    } else if mt.starts_with("video/") {
        FileCategory::Video
    } else {
        FileCategory::Other
    }
}

// ─── Core algorithm ───────────────────────────────────────────────────────────

/// MD5(id) → binary string → `1` → `*`, `0` → `:`.
///
/// This is the core encryption function matching the TS `generateEncryptedName`.
pub fn generate_encrypted_name(id: &str) -> String {
    let mut hasher = Md5::new();
    hasher.update(id.as_bytes());
    let hash = hasher.finalize();

    // Convert 16 bytes → 128-bit binary string, replacing 1 → '*', 0 → ':' in a single pass
    let mut result = String::with_capacity(128);
    for byte in hash.iter() {
        for i in (0..8).rev() {
            if (byte >> i) & 1 == 1 {
                result.push('*');
            } else {
                result.push(':');
            }
        }
    }

    result
}

/// Build the encrypted filename for a manifest item.
///
/// Port of TS `buildEncryptedFilename`.
pub fn build_encrypted_filename(id: &str, href: &str) -> String {
    let id_name = id.split('.').next().unwrap_or(id);
    let ext = href
        .rfind('.')
        .map(|i| &href[i + 1..])
        .unwrap_or("")
        .to_lowercase();

    let filename_without_ext = {
        let basename = href.rsplit('/').next().unwrap_or(href);
        basename
            .rfind('.')
            .map(|i| &basename[..i])
            .unwrap_or(basename)
    };

    let mut slim_suffix = "";
    let mut hash_input = id_name.to_string();
    if filename_without_ext.ends_with("slim") || id_name.ends_with("slim") {
        slim_suffix = "~slim";
        hash_input = hash_input
            .trim_end_matches("~slim")
            .trim_end_matches("-slim")
            .trim_end_matches("_slim")
            .trim_end_matches("slim")
            .to_string();
    }

    let encrypted = generate_encrypted_name(&hash_input);
    format!("_{encrypted}{slim_suffix}.{ext}")
}

/// Resolve a relative href against a base file path.
///
/// e.g. `resolve_book_path("../Images/cover.jpg", "OEBPS/Text/ch1.xhtml")`
///      → `"OEBPS/Images/cover.jpg"`
pub fn resolve_book_path(href: &str, base_path: &str) -> String {
    let parts: Vec<&str> = href.split(&['/', '\\'][..]).collect();
    let base_parts: Vec<&str> = base_path.split(&['/', '\\'][..]).collect();
    // Remove filename from base
    let base_dir = &base_parts[..base_parts.len().saturating_sub(1)];

    let mut back_steps = 0usize;
    let mut start = 0;
    for (i, part) in parts.iter().enumerate() {
        if *part == ".." {
            back_steps += 1;
            start = i + 1;
        } else if *part == "." {
            start = i + 1;
        } else {
            break;
        }
    }

    let remaining = &parts[start..];
    if back_steps >= base_dir.len() {
        return remaining.join("/");
    }

    let base_prefix = &base_dir[..base_dir.len() - back_steps];
    let mut result: Vec<&str> = base_prefix.to_vec();
    result.extend_from_slice(remaining);
    result.join("/")
}

/// Build a path mapping for all manifest items (excluding the TOC NCX).
///
/// Returns: `(path_map, href_to_new)` where:
/// - `path_map`: old bookpath → (new_basename, category)
/// - `href_to_new`: old OPF-relative href → new encrypted basename
pub fn build_path_map(
    manifest: &HashMap<String, ManifestItem>,
    toc_id: &str,
    opf_dir: &str,
) -> (HashMap<String, (String, FileCategory)>, HashMap<String, String>) {
    let mut path_map: HashMap<String, (String, FileCategory)> = HashMap::new();
    let mut href_to_new: HashMap<String, String> = HashMap::new();
    let mut used_names: HashMap<FileCategory, HashSet<String>> = HashMap::new();

    for (id, item) in manifest {
        if !toc_id.is_empty() && id == toc_id {
            continue;
        }

        let category = classify_item(item);
        let mut new_name = build_encrypted_filename(id, &item.href);

        // Ensure uniqueness within category
        let name_set = used_names.entry(category).or_default();
        if name_set.contains(&new_name) {
            let ext_idx = new_name.rfind('.').unwrap_or(new_name.len());
            let base = &new_name[..ext_idx];
            let ext = &new_name[ext_idx..];
            let mut counter = 1;
            loop {
                let candidate = format!("{base}_{counter}{ext}");
                if !name_set.contains(&candidate) {
                    new_name = candidate;
                    break;
                }
                counter += 1;
            }
        }
        name_set.insert(new_name.clone());

        let book_path = if opf_dir.is_empty() {
            item.href.clone()
        } else {
            format!("{}/{}", opf_dir, item.href)
        };
        path_map.insert(book_path, (new_name.clone(), category));
        href_to_new.insert(item.href.clone(), new_name);
    }

    (path_map, href_to_new)
}

// ─── Content rewriting helpers ────────────────────────────────────────────────

/// Rewrite href/src references in XHTML content.
pub fn rewrite_xhtml(
    text: &str,
    xhtml_book_path: &str,
    path_map: &HashMap<String, (String, FileCategory)>,
) -> String {
    let mut result = text.to_string();

    // Rewrite href="...", src="...", xlink:href="...", poster="..."
    for attr in &["href", "src", "xlink:href", "poster"] {
        result = rewrite_attr_references(&result, attr, xhtml_book_path, path_map);
    }

    // Rewrite url() in inline styles
    result = rewrite_url_references(&result, xhtml_book_path, path_map);

    result
}

/// Rewrite url() references in CSS content.
pub fn rewrite_css(
    css: &str,
    css_book_path: &str,
    path_map: &HashMap<String, (String, FileCategory)>,
) -> String {
    let mut result = css.to_string();

    // Rewrite url()
    result = rewrite_url_references(&result, css_book_path, path_map);

    result
}

/// Rewrite attribute references (href, src, etc.) in HTML/XHTML.
fn rewrite_attr_references(
    text: &str,
    attr_name: &str,
    base_path: &str,
    path_map: &HashMap<String, (String, FileCategory)>,
) -> String {
    // Simple regex-like approach: find attr="value" patterns
    let mut result = String::with_capacity(text.len());
    let search_pattern = format!("{attr_name}=\"");
    let mut pos = 0;

    while pos < text.len() {
        if let Some(start) = text[pos..].find(&search_pattern) {
            let abs_start = pos + start;
            let value_start = abs_start + search_pattern.len();

            if let Some(end_quote) = text[value_start..].find('"') {
                let value_end = value_start + end_quote;
                let href = &text[value_start..value_end];

                // Push everything up to the value
                result.push_str(&text[pos..value_start]);

                // Try to rewrite
                if let Some(new_href) = try_rewrite_href(href, base_path, path_map) {
                    result.push_str(&new_href);
                } else {
                    result.push_str(href);
                }

                pos = value_end;
            } else {
                result.push_str(&text[pos..abs_start + search_pattern.len()]);
                pos = abs_start + search_pattern.len();
            }
        } else {
            result.push_str(&text[pos..]);
            break;
        }
    }

    result
}

/// Rewrite `url()` references in CSS/inline styles.
fn rewrite_url_references(
    text: &str,
    base_path: &str,
    path_map: &HashMap<String, (String, FileCategory)>,
) -> String {
    let mut result = String::with_capacity(text.len());
    let mut pos = 0;

    while pos < text.len() {
        if let Some(url_start) = text[pos..].find("url(") {
            let abs_start = pos + url_start;
            let value_start = abs_start + 4;

            // Find closing )
            if let Some(paren_end) = text[value_start..].find(')') {
                let value_end = value_start + paren_end;
                let raw_url = text[value_start..value_end].trim();
                let url = raw_url.trim_matches(|c| c == '"' || c == '\'');

                result.push_str(&text[pos..abs_start]);
                result.push_str("url(");

                if let Some(new_href) = try_rewrite_href(url, base_path, path_map) {
                    result.push_str(&new_href);
                } else {
                    result.push_str(raw_url);
                }

                result.push(')');
                pos = value_end + 1;
            } else {
                result.push_str(&text[pos..abs_start + 4]);
                pos = abs_start + 4;
            }
        } else {
            result.push_str(&text[pos..]);
            break;
        }
    }

    result
}

/// Try to rewrite a single href to the new encrypted path.
fn try_rewrite_href(
    href: &str,
    base_path: &str,
    path_map: &HashMap<String, (String, FileCategory)>,
) -> Option<String> {
    let decoded = urlencoding_decode(href);
    let trimmed = decoded.trim();

    if trimmed.is_empty()
        || trimmed.starts_with("http://")
        || trimmed.starts_with("https://")
        || trimmed.starts_with("data:")
        || trimmed.starts_with("mailto:")
    {
        return None;
    }

    // Split off fragment
    let (href_base, fragment) = if let Some(hash_idx) = trimmed.find('#') {
        (&trimmed[..hash_idx], &trimmed[hash_idx..])
    } else {
        (trimmed, "")
    };

    if href_base.is_empty() {
        return None;
    }

    let book_path = resolve_book_path(href_base, base_path);
    let (new_basename, category) = path_map.get(&book_path)?;

    let dir = category.dir_name();
    Some(format!("../{dir}/{new_basename}{fragment}"))
}

/// Percent-decode a URL-encoded string (RFC 3986).
fn urlencoding_decode(s: &str) -> String {
    let mut result = Vec::with_capacity(s.len());
    let bytes = s.as_bytes();
    let mut i = 0;
    while i < bytes.len() {
        if bytes[i] == b'%' && i + 2 < bytes.len() {
            if let (Some(hi), Some(lo)) = (
                hex_val(bytes[i + 1]),
                hex_val(bytes[i + 2]),
            ) {
                result.push(hi << 4 | lo);
                i += 3;
                continue;
            }
        }
        result.push(bytes[i]);
        i += 1;
    }
    String::from_utf8(result).unwrap_or_else(|_| s.to_string())
}

fn hex_val(b: u8) -> Option<u8> {
    match b {
        b'0'..=b'9' => Some(b - b'0'),
        b'a'..=b'f' => Some(b - b'a' + 10),
        b'A'..=b'F' => Some(b - b'A' + 10),
        _ => None,
    }
}

// ─── Tests ────────────────────────────────────────────────────────────────────

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_generate_encrypted_name() {
        let name = generate_encrypted_name("chapter1");
        // Should be 128 chars of * and :
        assert_eq!(name.len(), 128);
        assert!(name.chars().all(|c| c == '*' || c == ':'));
    }

    #[test]
    fn test_generate_encrypted_name_deterministic() {
        let a = generate_encrypted_name("test_id");
        let b = generate_encrypted_name("test_id");
        assert_eq!(a, b);
    }

    #[test]
    fn test_generate_encrypted_name_different_inputs() {
        let a = generate_encrypted_name("chapter1");
        let b = generate_encrypted_name("chapter2");
        assert_ne!(a, b);
    }

    #[test]
    fn test_build_encrypted_filename() {
        let name = build_encrypted_filename("chapter1.xhtml", "Text/chapter1.xhtml");
        assert!(name.starts_with('_'));
        assert!(name.ends_with(".xhtml"));
        assert_eq!(name.len(), 1 + 128 + 6); // _ + 128chars + .xhtml
    }

    #[test]
    fn test_build_encrypted_filename_slim() {
        let name = build_encrypted_filename("chapter1slim.css", "Styles/chapter1slim.css");
        assert!(name.contains("~slim"));
        assert!(name.ends_with(".css"));
    }

    #[test]
    fn test_resolve_book_path() {
        assert_eq!(
            resolve_book_path("../Images/cover.jpg", "OEBPS/Text/ch1.xhtml"),
            "OEBPS/Images/cover.jpg"
        );
    }

    #[test]
    fn test_resolve_book_path_same_dir() {
        assert_eq!(
            resolve_book_path("style.css", "OEBPS/Text/ch1.xhtml"),
            "OEBPS/Text/style.css"
        );
    }

    #[test]
    fn test_resolve_book_path_up_two_levels() {
        assert_eq!(
            resolve_book_path("../../Images/img.png", "OEBPS/Text/sub/ch1.xhtml"),
            "OEBPS/Images/img.png"
        );
    }

    #[test]
    fn test_classify_item() {
        let text_item = ManifestItem {
            id: "ch1".into(),
            href: "Text/ch1.xhtml".into(),
            media_type: "application/xhtml+xml".into(),
            properties: None,
        };
        assert_eq!(classify_item(&text_item), FileCategory::Text);

        let css_item = ManifestItem {
            id: "style".into(),
            href: "Styles/main.css".into(),
            media_type: "text/css".into(),
            properties: None,
        };
        assert_eq!(classify_item(&css_item), FileCategory::Css);

        let img_item = ManifestItem {
            id: "cover".into(),
            href: "Images/cover.jpg".into(),
            media_type: "image/jpeg".into(),
            properties: None,
        };
        assert_eq!(classify_item(&img_item), FileCategory::Image);

        let font_item = ManifestItem {
            id: "font1".into(),
            href: "Fonts/myfont.ttf".into(),
            media_type: "font/ttf".into(),
            properties: None,
        };
        assert_eq!(classify_item(&font_item), FileCategory::Font);
    }

    #[test]
    fn test_rewrite_xhtml_href() {
        let mut path_map = HashMap::new();
        path_map.insert(
            "OEBPS/Styles/main.css".to_string(),
            ("_encrypted.css".to_string(), FileCategory::Css),
        );

        let xhtml = r#"<link href="../Styles/main.css" type="text/css"/>"#;
        let result = rewrite_xhtml(xhtml, "OEBPS/Text/ch1.xhtml", &path_map);
        assert!(result.contains("../Styles/_encrypted.css"));
    }

    #[test]
    fn test_rewrite_xhtml_src() {
        let mut path_map = HashMap::new();
        path_map.insert(
            "OEBPS/Images/cover.jpg".to_string(),
            ("_encrypted.jpg".to_string(), FileCategory::Image),
        );

        let xhtml = r#"<img src="../Images/cover.jpg" alt="cover"/>"#;
        let result = rewrite_xhtml(xhtml, "OEBPS/Text/ch1.xhtml", &path_map);
        assert!(result.contains("../Images/_encrypted.jpg"));
    }

    #[test]
    fn test_rewrite_ignores_external_urls() {
        let path_map = HashMap::new();
        let xhtml = r#"<a href="https://example.com">link</a>"#;
        let result = rewrite_xhtml(xhtml, "OEBPS/Text/ch1.xhtml", &path_map);
        assert_eq!(result, xhtml);
    }
}
